File: src/main/java/org/embulk/input/randomj/RandomjInputPlugin.java
package org.embulk.input.randomj;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.embulk.config.ConfigDiff;
import org.embulk.config.ConfigSource;
import org.embulk.config.TaskReport;
import org.embulk.config.TaskSource;
import org.embulk.spi.Column;
import org.embulk.spi.ColumnConfig;
import org.embulk.spi.Exec;
import org.embulk.spi.InputPlugin;
import org.embulk.spi.PageBuilder;
import org.embulk.spi.PageOutput;
import org.embulk.spi.Schema;
import org.embulk.spi.SchemaConfig;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;

public class RandomjInputPlugin
        implements InputPlugin
{
    @Override
    public ConfigDiff transaction(ConfigSource config,
            InputPlugin.Control control)
    {
        PluginTask task = config.loadConfig(PluginTask.class);

        Schema schema = task.getSchema().toSchema();
        int taskCount = task.getThreads();  // number of run() method calls

        return resume(task.dump(), schema, taskCount, control);
    }

    @Override
    public ConfigDiff resume(TaskSource taskSource,
            Schema schema, int taskCount,
            InputPlugin.Control control)
    {
        control.run(taskSource, schema, taskCount);
        return Exec.newConfigDiff();
    }

    @Override
    public void cleanup(TaskSource taskSource,
            Schema schema, int taskCount,
            List<TaskReport> successTaskReports)
    {
        // throw new UnsupportedOperationException("randomj input plugin does not support cleanup");
    }

    @Override
    public TaskReport run(TaskSource taskSource,
            Schema schema, int taskIndex,
            PageOutput output)
    {
        PluginTask task = taskSource.loadTask(PluginTask.class);
        Integer rows = task.getRows();
        final HashMap<Column, Map<String, Integer>> columnOptions = getColumnOptions(task);
        final HashMap<Column, List<JsonNode>> columnSchemas = getColumnSchemas(task);
        try (PageBuilder pagebuilder =
                new PageBuilder(Exec.getBufferAllocator(), schema, output)) {
            IntStream.rangeClosed(
                    taskIndex * rows + 1,
                    taskIndex * rows + rows
            ).boxed().forEach(rowNumber -> {
                RandomjColumnVisitor visitor = new RandomjColumnVisitor(pagebuilder, task, rowNumber, columnOptions, columnSchemas);
                schema.visitColumns(visitor);
                pagebuilder.addRecord();
            });
            pagebuilder.finish();
        }

        TaskReport taskReport = Exec.newTaskReport();
        taskReport.set("columns", schema.size());
        taskReport.set("rows", rows);
        return taskReport;
    }

    HashMap<Column, Map<String, Integer>> getColumnOptions(PluginTask task)
    {
        SchemaConfig schemaConfig = task.getSchema();
        Schema schema = schemaConfig.toSchema();
        HashMap<Column, Map<String, Integer>> lengthMap = new HashMap<>();
        for (Column column : schema.getColumns()) {
            HashMap<String, Integer> miniMap = new HashMap<>();
            ColumnConfig c = schemaConfig.getColumn(column.getIndex());
            miniMap.put("length", c.getOption().get(Integer.class, "length", 0));
            miniMap.put("null_rate", c.getOption().get(Integer.class, "null_rate", 0));
            miniMap.put("max_value", c.getOption().get(Integer.class, "max_value", null));
            miniMap.put("min_value", c.getOption().get(Integer.class, "min_value", null));
            lengthMap.put(column, miniMap);
        }
        return lengthMap;
    }

    HashMap<Column, List<JsonNode>> getColumnSchemas(PluginTask task)
    {
        SchemaConfig schemaConfig = task.getSchema();
        Schema schema = schemaConfig.toSchema();
        HashMap<Column, List<JsonNode>> schemaMap = new HashMap<>();
        for (Column column : schema.getColumns()) {
            String schemaString = schemaConfig
                    .getColumn(column.getIndex())
                    .getOption().get(String.class, "schema", "");
            if (!schemaString.isEmpty()) {
                try {
                    List<JsonNode> jsonNodes = new ObjectMapper().readValue(schemaString, new TypeReference<List<JsonNode>>() {});
                    schemaMap.put(column, jsonNodes);
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return schemaMap;
    }

    @Override
    public ConfigDiff guess(ConfigSource config)
    {
        return Exec.newConfigDiff();
    }
}

File: src/main/scala/org/embulk/input/randomj/RandomjInputPlugin.scala
package org.embulk.input.randomj

import java.time.{LocalDateTime, ZoneId}
import java.util
import java.util.Random
import java.util.stream.IntStream

import org.apache.commons.text.{CharacterPredicates, RandomStringGenerator}
import org.embulk.config.{ConfigDiff, ConfigSource, TaskReport, TaskSource}
import org.embulk.spi.{Exec, InputPlugin, PageBuilder, PageOutput, Schema}
import org.embulk.spi.time.Timestamp

object RandomjInputPlugin {
}

class RandomjInputPlugin extends InputPlugin {
  final private val generator = new RandomStringGenerator.Builder().withinRange('0', 'z').filteredBy(CharacterPredicates.LETTERS, CharacterPredicates.DIGITS).build

  override def transaction(config: ConfigSource, control: InputPlugin.Control): ConfigDiff = {
    val task = config.loadConfig(classOf[PluginTask])
    val schema = task.getSchema.toSchema
    val taskCount = task.getThreads // number of run() method calls
    resume(task.dump, schema, taskCount, control)
  }

  override def resume(taskSource: TaskSource, schema: Schema, taskCount: Int, control: InputPlugin.Control): ConfigDiff = {
    control.run(taskSource, schema, taskCount)
    Exec.newConfigDiff
  }

  override def cleanup(taskSource: TaskSource, schema: Schema, taskCount: Int, successTaskReports: util.List[TaskReport]): Unit = {
  }

  override def run(taskSource: TaskSource, schema: Schema, taskIndex: Int, output: PageOutput): TaskReport = {
    val task = taskSource.loadTask(classOf[PluginTask])
    val rows = task.getRows.asInstanceOf[Integer]
    try {
      val pagebuilder = new PageBuilder(Exec.getBufferAllocator, schema, output)
      try {
        val rnd = new Random
        java.util.stream.IntStream.rangeClosed(taskIndex * rows + 1, taskIndex * rows + rows).boxed.forEach((rowNumber: Integer) => {
          def foo(rowNumber: Integer) = {
            var i = 0
            while ( {
              i < schema.size
            }) {
              val column = schema.getColumn(i)
              val `type` = column.getType
              `type`.getName match {
                case "long" =>
                  val pk = task.getPrimaryKey
                  if (column.getName == pk) pagebuilder.setLong(i, rowNumber)
                  else pagebuilder.setLong(i, rnd.nextInt(10000))
                  break //todo: break is not supported
                case "double" =>
                  pagebuilder.setDouble(i, rnd.nextDouble * 10000)
                  break //todo: break is not supported
                case "boolean" =>
                  if (Math.random < 0.5) pagebuilder.setBoolean(i, false)
                  else pagebuilder.setBoolean(i, true)
                  break //todo: break is not supported
                case "string" =>
                  pagebuilder.setString(i, generator.generate(32))
                  break //todo: break is not supported
                case "timestamp" =>
                  val zoneId = ZoneId.systemDefault
                  val randd = Math.random
                  val randomDate = LocalDateTime.now.plusDays((randd * 100).toLong).plusSeconds((randd * 1000000).toLong)
                  val timestamp = Timestamp.ofEpochSecond(randomDate.atZone(zoneId).toEpochSecond)
                  pagebuilder.setTimestamp(column, timestamp)
                  break //todo: break is not supported
                case _ =>
                  System.out.println("Unsupported type")
                  break //todo: break is not supported
              }
              {
                i += 1
                i - 1
              }
            }
            pagebuilder.addRecord()
          }

          foo(rowNumber)
        })
        pagebuilder.finish()
      } finally if (pagebuilder != null) pagebuilder.close()
    }
    val taskReport = Exec.newTaskReport
    taskReport.set("columns", schema.size)
    taskReport.set("rows", rows)
    taskReport
  }

  override def guess(config: ConfigSource): ConfigDiff = Exec.newConfigDiff
}